Add support for lazy loading images

sinofp opened the pull request:
**IMPORTANT: Please do not create a Pull Request adding a new feature without discussing it first.**

The place to discuss new features is the forum: <https://zola.discourse.group/>
If you want to add a new feature, please open a thread there first in the feature requests section.

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

---

Recently I learnt `loading="lazy" decoding="async"` and wanted to try them in my Zola-powered blog, then I found issue #2055 requesting `loading="lazy"`.
Given that issue has been marked by “help wanted”, I decided to give it a try, hence this PR.

This PR adds an option `lazy_async_image` to set `loading="lazy" decoding="async"` for images.
The `loading="lazy"` part is discussed in the issue, I can delete `decoding="async"` part if that’s too far.

The alt text in pulldown-cmark is enclosed within `Start(Image)` and `End(Image)`, and can contain styles (<https://github.com/raphlinus/pulldown-cmark/issues/394#issuecomment-527197423>).
I think most people will only use plain text in alt text so I didn’t bother cleaning the styles because “Having something simple and easy to use for 90% of the usecases is more interesting than covering 100% usecases after sacrificing simplicity.”
That said, there might be people want to use styles, links or even inline HTML for alt text, so the documentation includes a note on the limitation of the `lazy_async_image` option.

I added a small test to demonstrate the difference between the output of pulldown-cmark and mine. I also tested the program against my blog.

sinofp committed:
Add optional decoding="async" loading="lazy" for img

In theory, they can make the page load faster and show content faster.

There’s one problem: CommonMark allows arbitrary inline elements in alt text.
If I want to get the correct alt text, I need to match every inline event.

I think most people will only use plain text, so I only match Event::Text.

sinofp committed:
Add very basic test for img

This is the reason why we should use plain text when lazy_async_image is enabled.

sinofp committed:
Explain lazy_async_image in documentation

Keats commented on components/markdown/src/markdown.rs:
```diff
@@ -326,7 +329,13 @@ pub fn markdown_to_html(
             match event {
                 Event::Text(text) => {
-                    if let Some(ref mut _code_block) = code_block {
+                    if lazy_async_image && is_image_alt {
```
Can we not just push back the Text event in that case and let pulldown-cmark handle it?

sinofp replied:
Indeed, this addition is redundant. I have removed it.

Keats commented on components/markdown/src/markdown.rs:
```diff
@@ -387,13 +396,38 @@ pub fn markdown_to_html(
+                            cmark::escape::escape_href(&mut img_before_alt, &title)
+                                .expect("Could not write to buffer");
+                        }
+                        img_before_alt.write_str("\" alt=\"").expect("Could not write to buffer");
```
It's not certain there will be an alt right?

sinofp replied:
Yes, the user might not input any alt text, but we still need to emit an empty alt attribute for the screen reader to ignore the picture: https://www.boia.org/blog/images-that-dont-need-alternative-text-still-need-alt-attributes

Keats commented on components/markdown/tests/img.rs:
```diff
@@ -0,0 +1,48 @@
+    .unwrap()
+    .body;
+    assert_eq!(rendered, "<p><img src=\"https://example.com/abc.jpg\" alt=\"ha<a href=\"https://example.com\">ha</a>\" loading=\"lazy\" decoding=\"async\" /></p>\n");
+}
```
Can you add a test with an empty alt?

sinofp replied:
Done! I added it for both pulldown-cmark and mine.

Keats replied:
Sorry I forgot zola was already using [insta](https://crates.io/crates/cargo-insta) for snapshot testing. Can you check eg `can_render_basic_markdown` to see how it works? Otherwise I'll do it myself later

sinofp replied:
No worries, I want to try it, so I’ve converted the test.

Keats reviewed (changes requested):
That's really cool, I didn't know about `decoding="async"`

sinofp committed:
Add test with empty alt and special characters

I totaly forgot one can leave the alt text empty.
I thought I need to eliminate the alt attribute in that case,
but actually empty alt text is better than not having an alt attribute at all:
https://www.w3.org/TR/WCAG20-TECHS/H67.html
https://www.boia.org/blog/images-that-dont-need-alternative-text-still-need-alt-attributes
Thus I will leave the empty alt text.

Another test is added to ensure alt text is properly escaped.
I will remove the redundant escaping code after this commit.

sinofp committed:
Remove manually escaping alt text

After removing the if-else inside the arm of Event::Text(text),
the alt text is still escaped.
Indeed they are redundant.

sinofp committed:
Use insta for snapshot testing

`cargo insta review` looks cool!

I wanted to dedup the cases variable,
but my Rust skill is not good enough to declare a global vector.
